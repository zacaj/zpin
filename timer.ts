import { Utils, Opaque, assert } from './util';
import { Events } from './events';
import { StateEvent } from './state';
import { Log } from './log';
import { fork, settleForks } from './promises';

export type Time = Opaque<number, 'Time'>;

export class Timer {
    static startTime = new Date().getTime();

    static mockTime?: number;
    static get time(): Time {
        if (Utils.stateAccessRecorder) {
            Utils.stateAccessRecorder(Timer, 'time');
        }

        if (Timer.mockTime !== undefined) return Timer.mockTime as Time;
        
        return new Date().getTime() - Timer.startTime as Time;
    }

    static queue: TimerQueueEntry[] = [];

    private static addToQueue(entry: TimerQueueEntry) {
        let i = 0;
        for (; i<Timer.queue.length; i++) {
            if (Timer.queue[i].time >= entry.time) {
                break;
            }
        }
        
        Timer.queue.splice(i, 0, entry);
    }

    static schedule(func: TimerCallback, at: Time, context?: string): TimerQueueEntry {
        const entry = { func, time: at, context };
        Timer.addToQueue(entry);
        return entry;
    }

    static callIn(func: TimerCallback, ms: number, context?: string) {
        return Timer.schedule(func, time() + ms as Time, context);
    }

    static setInterval(func: TimerCallback, ms: number, context?: string) {
        const entry = Timer.callIn(func, ms, context);
        entry.repeat = ms as Time;
        return entry;
    }

    static cancel(entry: TimerQueueEntry) {
        Timer.queue.remove(entry);
    }

    static curTime = time();

    static async fireTimers(before = time(), updateMock = false) {
        while (true) {
            const entry = Timer.queue[0];
            if (entry && entry.time <= before) {
                Timer.cancel(entry);
                try {
                    if (updateMock)
                        Timer.mockTime = entry.time;
                    await entry.func(entry);
                    await settleForks();
                } catch (err) {
                    Log.error('console', 'error running entry %o: ', entry, err);
                    debugger;
                }
                if (entry.repeat) {
                    entry.time = time() + entry.repeat as Time;
                    Timer.addToQueue(entry);
                }
            } else {
                break;
            }
        }
        
        if (updateMock)
            Timer.mockTime = before;
    }

    static reset() {
        this.queue = [];
    }
}
setInterval(() => {
    if (Timer.mockTime) return;
    const tim = time();
    Events.fire(new StateEvent(Timer, 'time', tim, Timer.curTime));
    Timer.curTime = tim;
    fork(Timer.fireTimers(tim), 'timer tick');
}, 5);

export type TimerCallback = (entry: TimerQueueEntry) => Promise<any>|any;
export type TimerQueueEntry = {
    time: Time;
    func: TimerCallback;
    repeat?: Time;
    context?: string;
};

export function safeSetInterval(func: () => any, ms: number, context: string) {
    // setInterval(() => {
    //     if (Timer.mockTime) return 'skipped';
    //     return func();
    // }, ms);
    Timer.setInterval(func, ms, context);
}
export function safeSetTimeout(func: () => any, ms: number, context: string) {
    // assert(!Timer.mockTime);
    // setTimeout(func, ms);
    Timer.callIn(func, ms, context);
}
export function time() {
    return Timer.time;
}
export async function setTime(ms?: number) {
    const lastTime = Timer.mockTime;
    Timer.mockTime = ms;
    const newTime = time();
    if (ms === undefined && lastTime) {
        Timer.startTime -= lastTime - newTime;
        const newNewTime = time();
        assert(Math.abs(newNewTime - lastTime) < 2);
        Log.log('console', 'adjusted time to %i', newNewTime);
    }
    await Timer.fireTimers(newTime, !!ms);
    Events.fire(new StateEvent(Timer, 'time', ms as Time ?? time(), lastTime as Time));
    if (Timer.mockTime !== undefined)
        await settleForks();
}
export async function passTime(ms = 1) {
    assert(Timer.mockTime !== undefined);
    await setTime(Timer.mockTime! + ms);
}

export async function wait(ms: number, context?: string) {
    return new Promise(resolve => safeSetTimeout(resolve, ms, context ?? 'wait '+ms));
}