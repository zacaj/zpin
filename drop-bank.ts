import { MachineOutputs, MomentarySolenoid, SolenoidFireEvent, Machine, ImageOutputs, Image, Light, machine } from './machine';
import { Mode } from './mode';
import { Switch, SwitchEvent, onSwitchClose, onAnySwitchClose } from './switch-matrix';
import { Outputs, toggle, OwnOutputEvent, TreeOutputEvent } from './outputs';
import { Event, Events, EventCallback, EventTypePredicate } from './events';
import { getTypeIn } from './util';
import { time } from './timer';
import { KnockTarget } from './util-modes';
import { State } from './state';
import { Log } from './log';
import { Tree } from './tree';
import { playSound } from './sound';

export type Standup = [
    sw: Switch,
    light: Light,
];

export interface DropTarget {
    state: boolean;
    bank: DropBank;
    num: number;
    switch: Switch;
    image: Image;
}

export class DropBank extends Tree<MachineOutputs> {
    targets: DropTarget[] = [];

    constructor(
        machine: Machine,
        public coil: MomentarySolenoid,
        public switches: Switch[],
        nums: number[],
        images: (keyof ImageOutputs)[],
    ) {
        super();
        let i=0;
        for (const sw of switches) {
            const target: DropTarget = {
                state: sw.state,
                bank: this,
                num: nums[i],
                image: new Image(images[i]),
                switch: sw,
            };
            State.declare<DropTarget>(target, ['state']);
            this.targets.push(target);
            machine.allDropTargets.push(target);
            if (nums[i]<17) machine.dropTargets.push(target);
            i++;
        }
        machine.dropBanks.push(this);

        this.listen([onAnySwitchClose(...switches), () => !coil.val],
            e => {
                const i = switches.indexOf(e.sw);
                if (this.targets[i].state) {
                    Log.info('switch', 'drop switch %s detected, but was already down', e.sw.name);
                    return;
                }
                void playSound('flip card');

                this.targets[i].state = true;
                Events.fire(new DropDownEvent(this.targets[i]));
                Log.info('switch', 'drop switch %s down', e.sw.name);
                if (this.targets.every(t => t.state)) {
                    Events.fire(new DropBankCompleteEvent(this));
                    Log.info('switch', 'drop bank %s complete', this.coil.name);
                }
            });

        this.listen(e => e instanceof TreeOutputEvent && e.tree === machine && e.prop === coil.name && e.value === false && e.oldValue === true,
            () => {
                Log.info('switch', 'drop bank %s reset successfully', this.coil.name);
                switches.forEach((_, i) => this.targets[i].state = false);
                Events.fire(new DropBankResetEvent(this));
            });
    }

    get numDown(): number {
        return this.targets.filter(t => t.state).length;
    }

    allAreUp(): boolean {
        return this.targets.every(t => t.state);
    }

    allAreDown(): boolean {
        return this.targets.every(t => t.state);
    }

    onTargetDown(num?: number): EventTypePredicate<DropDownEvent> {
        return e => e instanceof DropDownEvent && e.target.bank === this && (num === undefined || this.targets.indexOf(e.target) === num);
    }

    onAllDown(): EventTypePredicate<DropBankCompleteEvent> {
        return e => e instanceof DropBankCompleteEvent && e.bank === this;
    }

    cleanLog() {
        return `DropBank ${this.coil?.name}`;
    }
}

export class DropBankResetter extends Tree<MachineOutputs> {
    constructor(
        public bank: DropBank,
    ) {
        super();

        this.out = new Outputs(this, {
            [bank.coil.name]: toggle({
                on: () => bank.targets.every(t => t.switch.onFor(250)),
                off: () => bank.targets.every(t => !t.switch.state),
            }),
        });
    }
}

export class DropBankCompleteEvent extends Event {
    constructor(
        public bank: DropBank,
    ) {
        super();
    }
}
export class DropBankResetEvent extends Event {
    constructor(
        public bank: DropBank,
    ) {
        super();
    }
}
export class DropDownEvent extends Event {
    constructor(
        public target: DropTarget,
    ) {
        super();
    }
}