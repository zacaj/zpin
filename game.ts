import fs from 'fs';
import { argv } from 'yargs';
import { AttractMode } from './attract';
import { Events } from './events';
import { addToScreen, alert, gfx, screen } from './gfx';
import { GameGfx } from './gfx/game';
import { checkForScores } from './highscore';
import { initMachine } from './init';
import { Log } from './log';
import { machine } from './machine';
import { Mode, Modes } from './mode';
import { Player } from './modes/player';
import { MPU } from './mpu';
import { Outputs } from './outputs';
import { playSound } from './sound';
import { State } from './state';
import { onClose, onSwitchClose, onSwitchOpen, Switch } from './switch-matrix';
import { time, Timer } from './timer';
import { TreeChangeEvent } from './tree';
import { assert, getFormattedTime } from './util';
import { ClearHoles } from './util-modes';

export class Game extends Mode {
    get nodes() {
        return [this.curPlayer, ...this.tempNodes].truthy();
    }

    players: Player[] = [];
    playerUp = 0;
    get curPlayer(): Player {
        return this.players[this.playerUp];
    }
    ballNum = 1;
    ballCount = 3;

    get ball() {
        return this.curPlayer.ball;
    }

    startTimestamp = getFormattedTime();

    totals: {[source: string]: {times: number; total: number; average: number}} = {};
    
    private constructor(
        public seed: string,
    ) {
        super(Modes.Game);
        Log.log(['console', 'game'], "start game with seed ", seed);
        // assert(machine.sTroughFull.state);
        State.declare<Game>(this, ['ballNum', 'playerUp', 'players']);

        this.out = new Outputs(this, {
            kickerEnable: true,
            // magnetPost: () => (machine.sShooterUpper.wasClosedWithin(1000) || 
            //         (machine.sLeftOrbit.wasClosedWithin(2000) && !machine.sShooterUpper.wasClosedWithin(1000) && machine.cRightGate.actual))
            //         && !machine.sShooterLower.wasClosedWithin(750),
            // upperMagnet: () => machine.sShooterUpper.wasClosedWithin(3000) && !machine.sShooterLower.wasClosedWithin(750) && !machine.sSpinner.wasClosedWithin(750),
        });

        this.listen(onSwitchClose(machine.sStartButton), e => Timer.callIn(() => {
            if (machine.sStartButton.state && machine.sStartButton.lastClosed === e.when) {
                Log.log(['game', 'console'], 'game force-ended');
                this.end();
            }
        }, 2000, 'start button hold'));
        this.listen([...onSwitchOpen(machine.sStartButton), () => time()-machine.sStartButton.lastClosed! < 500 && machine.sStartButton.lastClosed! > this.startTime], 'addPlayer');
        

        // this.listen(onSwitchClose(machine.sLeftInlane),
        //     () => this.addTemp(new KnockTarget()));

        this.listen(onClose(), (e) => this.curPlayer.addScore(this.scores.get(e.sw) ?? 0, e.sw.name));

        addToScreen(() => new GameGfx(this));
    }

    async onBallEnd() {
        const lastPlayer = this.curPlayer;

        if (this.playerUp+1 < this.players.length)
            this.playerUp++;
        else {
            this.playerUp = 0;
            this.ballNum++;
            if (this.ballNum > this.ballCount) {
                // alert('GAME OVER', 5000);
                await checkForScores(this);
                const totals = Object.keys(this.totals).map(source => ({
                    ...this.totals[source],
                    source,
                }));
                totals.sort((a, b) => b.total - a.total);

                fs.writeFileSync(`./scores/game-${this.startTimestamp}.json`, JSON.stringify({
                    players: this.players.map(p => ({ number: p.number, score: p.score })),
                    totals,
                }, undefined, 2));
                if (require.main === module) {
                    debugger;
                    process.exit(0);
                }
                else 
                    this.end();
            }
        }
        if (!this.ended) {
            Log.log('console', 'player %i starting ball %i', this.curPlayer.number, this.ballNum);
            Events.fire(new TreeChangeEvent(this));
            await this.curPlayer.startBall();
        }
    }

    end() {
        machine.attract = new AttractMode(this.players.map(p => [p.score, p.store.Poker!.bank]));
        machine.attract.started();
        machine.game = undefined;
        return super.end();
    }

    static async start(seed: string = argv.seed as string ?? 'pinball'): Promise<Game> {
        assert(!machine.game);
        if (gfx && !MPU.isLive) {
            machine.sTroughFull.changeState(true, 'fake');
        }
        if (!machine.sTroughFull.state) {
            const pop = alert('BALL MISSING', 0)[0];
            const clear = new ClearHoles();
            machine.addTemp(clear);
            await new Promise(resolve => Events.listen(() => {
                resolve();
                return 'remove';
            }, onSwitchClose(machine.sTroughFull)));
            pop.parent?.remove(pop);
            clear.end();
        }
        machine.attract?.end();
        const game = new Game(seed);
        machine.game = game;
        game.started();
        game.players.push(new Player(game, 1, game.seed));
        game.curPlayer.started();
        await game.curPlayer.startBall();
        return game;
    }

    addPlayer() {
        if (!machine.sShooterLane.state || this.ballNum > 1 || this.curPlayer.score>0) return;
        this.players.push(new Player(this, this.players.length+1, this.seed));
        alert(`PLAYER ${this.players.length} ADDED`);
        if (this.players.length <= 4) void playSound(`player ${this.players.length}`);
    }

    scores = new Map<Switch, number>([
        [machine.sLeftInlane,       1000],
        [machine.sLeftOutlane,      10000],
        [machine.sRightInlane,      1000],
        [machine.sRightOutlane,     10000],
        [machine.sMiniOut,          0],
        [machine.sOuthole,          0],
        [machine.sTroughFull,       0],
        [machine.sLeftSling,        10],
        [machine.sRightSling,       100],
        [machine.sMiniLeft,         25000],
        [machine.sMiniCenter,       25000],
        [machine.sMiniRight,        25000],
        [machine.sCenterLeft,       6000],
        [machine.sCenterCenter,     6000],
        [machine.sCenterRight,      6000],
        [machine.sLeft1,            4000],
        [machine.sLeft2,            4000],
        [machine.sLeft3,            4000],
        [machine.sLeft4,            4000],
        [machine.sRight1,           4000],
        [machine.sRight2,           4000],
        [machine.sRight3,           4000],
        [machine.sRight4,           4000],
        [machine.sRight5,           4000],
        [machine.sLeftBack1,        5000],
        [machine.sLeftBack2,        5000],
        [machine.sUpper3Left,       12000],
        [machine.sUpper3Center,     12000],
        [machine.sUpper3Right,      12000],
        [machine.sUpper2Left,       12000],
        [machine.sUpper2Right,      12000],
        [machine.sSingleStandup,    13000],
        [machine.sRampMini,         3000],
        [machine.sRampMiniOuter,    3000],
        [machine.sRampDown,         0],
        [machine.sUnderRamp,        0],
        [machine.sLeftOrbit,        5000],
        [machine.sSpinner,          0],
        [machine.sSpinnerMini,      3000],
        [machine.sUpperPopMini,     3000],
        [machine.sSidePopMini,      3000],
        [machine.sShooterUpper,     10],
        [machine.sShooterMagnet,    10],
        [machine.sShooterLane,      0],
        [machine.sShooterLower,     10],
        [machine.sBackLane,         1000],
        [machine.sPop,              7500],
        [machine.sUpperInlane,      1000],
        [machine.sUnderUpperFlipper,25000],
        [machine.sUpperSideTarget,  15000],
        [machine.sUpperEject,       1000],
        [machine.sUpperLane2,    1000],
        [machine.sUpperLane3,   1000],
        [machine.sUpperLane4,    1000],
        [machine.sRampMade,         5000],
    ]);
}

if (require.main === module) {
    void initMachine(true, true, true, false);
}
