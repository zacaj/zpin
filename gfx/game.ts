import { Group, Text } from "aminogfx-gl";
import { Game } from "../game";
import { gfx, makeText, Screen } from "../gfx";
import { onChange } from "../state";

export class GameGfx extends Group {
    score = makeText('00', 60, 'center', 'top');
    ball = makeText('FREE PLAY', 40, 'right', 'top');
    constructor(
        public game: Game,
    ) {
        super(gfx);
        this.z(-1000);

        this.add(this.score.y(-Screen.h/2));
        game.listen(onChange(game.players[0], 'score'), (e) => this.score.text(e.value!.toFixed(0)));

        this.add(this.ball.x(Screen.w/2).y(-Screen.h/2));
        game.listen(onChange(game, 'ballNum'), (e) => this.ball.text('BALL '+e.value!.toFixed(0)));
    }
}