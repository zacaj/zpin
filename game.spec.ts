import { expectMachineOutputs, machine } from './machine'
import { Game } from './game';
import { getTypeIn } from './util';
import { DropBank } from './drop-bank';
import { passTime } from './timer';

describe('game', () => {
    test('resets drops', async () => {
        expectMachineOutputs('rampUp', 'upper3', 'centerBank', 'leftBank', 'rightBank', 'upper2', 'miniBank');
        const game = Game.start();
        for (const bank of getTypeIn<DropBank>(machine, DropBank)) {
            for (const sw of bank.targets.map(t => t.switch)) {
                sw.state = true;
            }
            await passTime(255);
            expect(bank.coil.set).toHaveBeenCalled();
        }
    });
});