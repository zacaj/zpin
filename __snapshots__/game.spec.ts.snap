// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game close diverter on orbit 1`] = `
Object {
  "outputs": Object {
    "centerBank": false,
    "getSkillshot": [Function],
    "iCenter1": "AD",
    "iCenter2": "KC",
    "iCenter3": "KS",
    "iLeft1": "8C",
    "iLeft2": "10S",
    "iLeft3": "JH",
    "iLeft4": "8S",
    "iMini1": "AH",
    "iMini2": "9S",
    "iMini3": "AC",
    "iRight1": "9C",
    "iRight2": "AS",
    "iRight3": "KH",
    "iRight4": "3D",
    "iRight5": "KD",
    "iSS1": "",
    "iSS2": "",
    "iSS3": "",
    "iSS4": "",
    "iSS5": "",
    "iSS6": "",
    "iSpinner": "",
    "iUpper21": "10H",
    "iUpper22": "QD",
    "iUpper31": "8D",
    "iUpper32": "10D",
    "iUpper33": "QS",
    "ignoreSkillsot": Set {},
    "kickerEnable": true,
    "lEjectArrow": Array [],
    "lEjectShowCards": Array [],
    "lEjectStartMode": Array [],
    "lLaneUpper1": Array [
      "Orange",
    ],
    "lLaneUpper2": Array [
      "Orange",
    ],
    "lLaneUpper3": Array [
      "Orange",
    ],
    "lLaneUpper4": Array [
      "Orange",
    ],
    "lLeftArrow": Array [],
    "lLowerRamp": Array [],
    "lMagnet1": Array [
      "Green",
    ],
    "lMagnet2": Array [
      "Green",
    ],
    "lMagnet3": Array [],
    "lMagnet4": Array [],
    "lMainTargetArrow": Array [],
    "lMiniReady": Array [
      "Red",
    ],
    "lPopperStatus": Array [
      "Green",
    ],
    "lPower1": Array [
      "Orange",
    ],
    "lPower2": Array [
      "Orange",
    ],
    "lPower3": Array [],
    "lPower4": Array [],
    "lRampArrow": Array [],
    "lRampMini": Array [
      "Orange",
    ],
    "lRampShowCards": Array [],
    "lRampStartMb": Array [],
    "lShooterLaneArrow": Array [],
    "lShooterShowCards": Array [],
    "lShooterStartHand": Array [],
    "lSideShotArrow": Array [],
    "lSideTargetArrow": Array [],
    "lSpinnerArrow": Array [],
    "lUpperLaneArrow": Array [],
    "lUpperTargetArrow": Array [],
    "leftBank": false,
    "leftGate": false,
    "leftMagnet": false,
    "lockPost": false,
    "magnetPost": false,
    "miniBank": false,
    "miniDiverter": false,
    "miniEject": false,
    "miniFlipperEnable": false,
    "outhole": false,
    "popper": false,
    "rampUp": false,
    "realRightBank": false,
    "right1": false,
    "right2": false,
    "right3": false,
    "right4": false,
    "right5": false,
    "rightBank": false,
    "rightGate": true,
    "shooterDiverter": true,
    "spinnerValue": undefined,
    "troughRelease": false,
    "upper2": false,
    "upper3": false,
    "upperEject": false,
    "upperMagnet": false,
  },
  "state": Object {
    "_name": "Machine10",
    "children": Array [
      Object {
        "_name": "DropBank11",
        "children": Array [],
        "name": "DropBank11",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank12",
        "children": Array [],
        "name": "DropBank12",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank13",
        "children": Array [],
        "name": "DropBank13",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank14",
        "children": Array [],
        "name": "DropBank14",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank15",
        "children": Array [],
        "name": "DropBank15",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "DropBank16",
        "children": Array [],
        "name": "DropBank16",
        "ownOuts": undefined,
        "state": Object {},
      },
      Object {
        "_name": "Game20",
        "children": Array [
          Object {
            "_name": "Player21",
            "children": Array [
              Object {
                "_name": "ClearHoles22",
                "children": Array [],
                "name": "ClearHoles22",
                "ownOuts": Object {
                  "miniEject": false,
                  "outhole": false,
                  "upperEject": false,
                },
                "state": Object {},
              },
              Object {
                "_name": "Ball26",
                "children": Array [
                  Object {
                    "_name": "ResetAnyDropOnComplete27",
                    "children": Array [
                      Object {
                        "_name": "DropBankResetter28",
                        "children": Array [],
                        "name": "DropBankResetter28",
                        "ownOuts": Object {
                          "upper3": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter29",
                        "children": Array [],
                        "name": "DropBankResetter29",
                        "ownOuts": Object {
                          "upper2": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter30",
                        "children": Array [],
                        "name": "DropBankResetter30",
                        "ownOuts": Object {
                          "centerBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter31",
                        "children": Array [],
                        "name": "DropBankResetter31",
                        "ownOuts": Object {
                          "miniBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter32",
                        "children": Array [],
                        "name": "DropBankResetter32",
                        "ownOuts": Object {
                          "leftBank": false,
                        },
                        "state": Object {},
                      },
                      Object {
                        "_name": "DropBankResetter33",
                        "children": Array [],
                        "name": "DropBankResetter33",
                        "ownOuts": Object {
                          "rightBank": false,
                        },
                        "state": Object {},
                      },
                    ],
                    "name": "ResetAnyDropOnComplete27",
                    "ownOuts": undefined,
                    "state": Object {},
                  },
                ],
                "name": "Ball26",
                "ownOuts": Object {},
                "state": Object {
                  "skillshot": undefined,
                  "tilted": false,
                },
              },
              Object {
                "_name": "Spinner23",
                "children": Array [],
                "name": "Spinner23",
                "ownOuts": Object {
                  "iSpinner": undefined,
                  "leftGate": false,
                },
                "state": Object {
                  "comboMult": 1,
                  "rounds": 0,
                  "score": 10,
                },
              },
              Object {
                "_name": "LeftOrbit24",
                "children": Array [],
                "name": "LeftOrbit24",
                "ownOuts": Object {
                  "rightGate": true,
                },
                "state": Object {
                  "comboMult": 1,
                  "rounds": 1,
                  "score": 20000,
                },
              },
              Object {
                "_name": "Poker34",
                "children": Array [],
                "name": "Poker34",
                "ownOuts": Object {
                  "getSkillshot": [Function],
                  "iCenter1": "AD",
                  "iCenter2": "KC",
                  "iCenter3": "KS",
                  "iLeft1": "8C",
                  "iLeft2": "10S",
                  "iLeft3": "JH",
                  "iLeft4": "8S",
                  "iMini1": "AH",
                  "iMini2": "9S",
                  "iMini3": "AC",
                  "iRight1": "9C",
                  "iRight2": "AS",
                  "iRight3": "KH",
                  "iRight4": "3D",
                  "iRight5": "KD",
                  "iUpper21": "10H",
                  "iUpper22": "QD",
                  "iUpper31": "8D",
                  "iUpper32": "10D",
                  "iUpper33": "QS",
                  "lEjectShowCards": Array [],
                  "lRampShowCards": Array [],
                  "lShooterShowCards": Array [],
                  "lockPost": undefined,
                  "rampUp": undefined,
                  "shooterDiverter": true,
                  "upperEject": undefined,
                },
                "state": Object {
                  "bet": 100,
                  "closeShooter": false,
                  "dealerCardsUsed": Array [],
                  "dealerHand": Array [
                    Object {
                      "flipped": true,
                      "num": 3,
                      "suit": "h",
                    },
                    Object {
                      "flipped": true,
                      "num": 4,
                      "suit": "c",
                    },
                    null,
                    null,
                    null,
                    null,
                    null,
                  ],
                  "dealerHandDesc": undefined,
                  "newMbs": Array [],
                  "newModes": Array [],
                  "playerCardsUsed": Array [],
                  "playerHand": Array [
                    Object {
                      "num": 7,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "s",
                    },
                    null,
                    null,
                    null,
                    null,
                    null,
                  ],
                  "playerHandDesc": undefined,
                  "playerWins": undefined,
                  "pot": 400,
                  "slots": Array [
                    Object {
                      "num": 1,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "c",
                    },
                    Object {
                      "num": 13,
                      "suit": "s",
                    },
                    Object {
                      "num": 9,
                      "suit": "c",
                    },
                    Object {
                      "num": 1,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "h",
                    },
                    Object {
                      "num": 3,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "d",
                    },
                    Object {
                      "num": 8,
                      "suit": "c",
                    },
                    Object {
                      "num": 10,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "h",
                    },
                    Object {
                      "num": 8,
                      "suit": "s",
                    },
                    Object {
                      "num": 10,
                      "suit": "h",
                    },
                    Object {
                      "num": 12,
                      "suit": "d",
                    },
                    Object {
                      "num": 8,
                      "suit": "d",
                    },
                    Object {
                      "num": 10,
                      "suit": "d",
                    },
                    Object {
                      "num": 12,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "h",
                    },
                    Object {
                      "num": 9,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "c",
                    },
                  ],
                  "step": 2,
                },
              },
              Object {
                "_name": "PlayerOverrides25",
                "children": Array [],
                "name": "PlayerOverrides25",
                "ownOuts": Object {
                  "kickerEnable": undefined,
                  "leftGate": undefined,
                  "miniFlipperEnable": undefined,
                  "rampUp": undefined,
                  "rightGate": undefined,
                  "shooterDiverter": undefined,
                },
                "state": Object {},
              },
            ],
            "name": "Player21",
            "ownOuts": Object {
              "lEjectStartMode": Array [],
              "lLaneUpper1": Array [
                "Orange",
              ],
              "lLaneUpper2": Array [
                "Orange",
              ],
              "lLaneUpper3": Array [
                "Orange",
              ],
              "lLaneUpper4": Array [
                "Orange",
              ],
              "lMiniReady": Array [
                "Red",
              ],
              "lPopperStatus": Array [
                "Green",
              ],
              "lPower1": Array [
                "Orange",
              ],
              "lPower2": Array [
                "Orange",
              ],
              "lPower3": Array [],
              "lPower4": Array [],
              "lRampMini": Array [
                "Orange",
              ],
              "lRampStartMb": Array [],
              "lShooterStartHand": Array [],
              "leftMagnet": false,
              "rampUp": false,
              "shooterDiverter": undefined,
            },
            "state": Object {
              "_score": 12010,
              "ball": "Ball26",
              "chips": 2,
              "closeShooter": false,
              "focus": "Poker34",
              "laneChips": Array [
                true,
                true,
                true,
                true,
              ],
              "mbsQualified": Array [],
              "miniReady": false,
              "modesQualified": Array [],
              "rampUp": false,
            },
          },
        ],
        "name": "Game20",
        "ownOuts": Object {
          "kickerEnable": true,
        },
        "state": Object {
          "ballNum": 1,
          "playerUp": 0,
          "players": Array [
            "Player21",
          ],
        },
      },
      Object {
        "_name": "MachineOverrides17",
        "children": Array [
          Object {
            "_name": "EosPulse18",
            "children": Array [],
            "name": "EosPulse18",
            "ownOuts": undefined,
            "state": Object {},
          },
        ],
        "name": "MachineOverrides17",
        "ownOuts": Object {
          "rampUp": undefined,
          "realRightBank": false,
          "shooterDiverter": undefined,
        },
        "state": Object {},
      },
    ],
    "name": "Machine10",
    "ownOuts": Object {
      "centerBank": false,
      "getSkillshot": undefined,
      "iCenter1": "",
      "iCenter2": "",
      "iCenter3": "",
      "iLeft1": "",
      "iLeft2": "",
      "iLeft3": "",
      "iLeft4": "",
      "iMini1": "",
      "iMini2": "",
      "iMini3": "",
      "iRight1": "",
      "iRight2": "",
      "iRight3": "",
      "iRight4": "",
      "iRight5": "",
      "iSS1": "",
      "iSS2": "",
      "iSS3": "",
      "iSS4": "",
      "iSS5": "",
      "iSS6": "",
      "iSpinner": "",
      "iUpper21": "",
      "iUpper22": "",
      "iUpper31": "",
      "iUpper32": "",
      "iUpper33": "",
      "ignoreSkillsot": Set {},
      "kickerEnable": false,
      "lEjectArrow": Array [],
      "lEjectShowCards": Array [],
      "lEjectStartMode": Array [],
      "lLaneUpper1": Array [],
      "lLaneUpper2": Array [],
      "lLaneUpper3": Array [],
      "lLaneUpper4": Array [],
      "lLeftArrow": Array [],
      "lLowerRamp": Array [],
      "lMagnet1": Array [
        "Green",
      ],
      "lMagnet2": Array [
        "Green",
      ],
      "lMagnet3": Array [],
      "lMagnet4": Array [],
      "lMainTargetArrow": Array [],
      "lMiniReady": Array [
        "Red",
      ],
      "lPopperStatus": Array [],
      "lPower1": Array [],
      "lPower2": Array [],
      "lPower3": Array [],
      "lPower4": Array [],
      "lRampArrow": Array [],
      "lRampMini": Array [],
      "lRampShowCards": Array [],
      "lRampStartMb": Array [],
      "lShooterLaneArrow": Array [],
      "lShooterShowCards": Array [],
      "lShooterStartHand": Array [],
      "lSideShotArrow": Array [],
      "lSideTargetArrow": Array [],
      "lSpinnerArrow": Array [],
      "lUpperLaneArrow": Array [],
      "lUpperTargetArrow": Array [],
      "leftBank": false,
      "leftGate": false,
      "leftMagnet": false,
      "lockPost": false,
      "magnetPost": false,
      "miniBank": false,
      "miniDiverter": false,
      "miniEject": false,
      "miniFlipperEnable": false,
      "outhole": false,
      "popper": false,
      "rampUp": false,
      "realRightBank": false,
      "right1": false,
      "right2": false,
      "right3": false,
      "right4": false,
      "right5": false,
      "rightBank": false,
      "rightGate": false,
      "shooterDiverter": false,
      "spinnerValue": undefined,
      "troughRelease": false,
      "upper2": false,
      "upper3": false,
      "upperEject": false,
      "upperMagnet": false,
    },
    "state": Object {
      "ballsInTrough": 3,
      "ballsLocked": 0,
      "lastSwitchHit": Switch {
        "_state": false,
        "column": 5,
        "lastChange": 15763,
        "lastClosed": 15497,
        "lastOpened": 15763,
        "minOffTime": 10,
        "minOnTime": 1,
        "name": "right inlane",
        "row": 0,
      },
      "lockDown": false,
      "miniDown": false,
    },
  },
}
`;
